lib/memoize.js

```diff
@@ -1,86 +1,144 @@
+      } else {
+        if (node.referenceArgs === null) {
+          argsMap = node.referenceArgs = new WeakMap();
+        } else {
+          argsMap = node.referenceArgs;
+        }
+      }
+
+      if (argsMap.has(arg) === false) {
+        node = {
+          primitiveResults: null,
+          referenceResults: null,
+          primitiveArgs: null,
+          referenceArgs: null
+        };
+        argsMap.set(arg, node);
+      } else {
+        node = argsMap.get(arg);
+      }
+    }
+
+    let result;
+
+    const arg = arguments[last];
+    const primitive =
+      arg === null || (typeof arg !== "object" && typeof arg !== "function");
+
+    let resultsMap;
 
-MapTree.prototype.clear = function clear() {
-  if (arguments.length === 0) {
-    this.childBranches = new WeakMap();
-    this.primitiveKeys.clear();
-    this.hasValue = false;
-    this.value = undefined;
-  } else if (arguments.length === 1) {
-    var key = arguments[0];
-    if (isPrimitive(key)) {
-      var keyObject = this.primitiveKeys.get(key);
-      if (keyObject) {
-        this.childBranches.delete(keyObject);
-        this.primitiveKeys.delete(key);
+    if (primitive) {
+      if (node.primitiveResults === null) {
+        resultsMap = node.primitiveResults = new Map();
+      } else {
+        resultsMap = node.primitiveResults;
       }
     } else {
-      this.childBranches.delete(key);
+      if (node.referenceResults === null) {
+        resultsMap = node.referenceResults = new WeakMap();
+      } else {
+        resultsMap = node.referenceResults;
+      }
     }
-  } else {
-    var childKey = arguments[0];
-    if (this.has(childKey)) {
-      var childBranch = this.get(childKey);
-      childBranch.clear.apply(childBranch, Array.prototype.slice.call(arguments, 1));
+
+    if (resultsMap.has(arg)) {
+      result = resultsMap.get(arg);
+    } else {
+      result = fn.apply(this, arguments);
+      resultsMap.set(arg, result);
     }
-  }
-};
 
-module.exports = function memoize(fn) {
-  var argsTree = new MapTree();
-
-  function memoized() {
-    var args = Array.prototype.slice.call(arguments);
-    var argNode = args.reduce(function getBranch(parentBranch, arg) {
-      return parentBranch.resolveBranch(arg);
-    }, argsTree);
-    if (argNode.hasValue) { return argNode.value; }
-    var value = fn.apply(null, args);
-    return argNode.setValue(value);
-  }
-
-  memoized.clear = argsTree.clear.bind(argsTree);
+    return result;
+  };
+
+  memoized.clear = function clear() {
+    const last = arguments.length - 1;
+    if (last === -1) {
+      reset();
+      return;
+    }
+
+    let node = root;
+
+    for (let i = 0; i < last; i += 1) {
+      const arg = arguments[i];
+      const primitive =
+        arg === null || (typeof arg !== "object" && typeof arg !== "function");
+      const argsMap = primitive ? node.primitiveArgs : node.referenceArgs;
+      if (argsMap === null) {
+        return;
+      }
+      node = argsMap.get(arg);
```

Comment:
Missing null check, if you call `memoized("foo", "bar", "baz")` and then `memoized.clear("fu", "bar")`, it throws null reference on `node` in the following code

Note: `clear` call has to have 2+ arguments and one of them expect the last must be on a non-memoized path